Templatize `_d_arraysetcapacity` hook

Albert24GG opened the pull request:
This replaces the `_d_arraysetcapacity` runtime hook from `rt/lifetime.d` with a templated version, now located in `core/internal/array/capacity.d`.

All the ***TypeInfo*** instances have been eliminated, except for the one pass as an argument to ```GC.malloc```.

The only walkaround that I needed to do is to use a "fake" pure & nothrow wrapper over the templated hook, since the `reserve` and `capacity` functions both are pure & nothrow, but the `__doPostblit` function can sometimes invoke impure postblits, depending on the type **T** passed. If you have any suggestions regarding this issue, please let me know!

Albert24GG committed:
First try for `_d_arraysetcapacity` templatization

Albert24GG committed:
Merge branch 'dlang:master' into arraysetcapacity-hook-template

Albert24GG committed:
Fix compilation issues with phobos

Albert24GG committed:
Templatize `__doPostblit` and `__typeAttrs` to avoid undefined refs

Albert24GG committed:
Remove `TypeInfo` dependencies from `__typeAttrs`

Albert24GG committed:
Remove redundant `TypeInfo` uses from `_d_arraysetcapacity`

Albert24GG committed:
Remove `_d_arraysetcapacity` from rt/lifetime.d

Albert24GG committed:
Use a fake pure wrapper over to avoid issues when postblit is impure

thewilsonator commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,182 @@
+
+Pass in a requested capacity of 0 to get the current capacity.
+
+Params:
+    ti = type info of element type
+    newcapacity = requested new capacity
+    p = pointer to array to set. Its `length` is left unchanged.
+
+Returns: the number of elements that can actually be stored once the resizing is done
+*/
+size_t _d_arraysetcapacityPureNothrow(T)(size_t newcapacity, void[]* p) pure nothrow @weak @trusted
```
```
 import/core/internal/array/capacity.d(79): Warning: Ddoc: function declaration has no parameter 'ti'
size_t _d_arraysetcapacityPureNothrow(T)(size_t newcapacity, void[]* p) pure nothrow @weak @trusted
       ^
import/core/internal/array/capacity.d(79): Warning: Ddoc: parameter count mismatch, expected 2, got 3
size_t _d_arraysetcapacityPureNothrow(T)(size_t newcapacity, void[]* p) pure nothrow @weak @trusted
```

Albert24GG committed:
docs: Remove ti parameter from `_d_arraysetcapacity` docs

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,181 @@
+private uint __typeAttrs(T)(void *copyAttrsFrom = null)
+{
+    import core.internal.traits : hasElaborateDestructor, hasIndirections;
+    import core.memory : GC;
+
+    alias BlkAttr = GC.BlkAttr;
+
+    if (copyAttrsFrom)
+    {
+        // try to copy attrs from the given block
+        auto info = GC.query(copyAttrsFrom);
+        if (info.base)
+            return info.attr;
+    }
+
+    uint attrs = 0;
+    static if (hasIndirections!T)
+        attrs |= BlkAttr.NO_SCAN;
+
+    static if (hasElaborateDestructor!T)
+        attrs |= BlkAttr.FINALIZE;
+
+    return attrs;
+}
+
+private void __doPostblit(T)(T[] arr)
+{
+    // infer static postblit type, run postblit if any
+    static if (__traits(hasPostblit, T))
+    {
+        static if (__traits(isStaticArray, T) && is(T : E[], E))
+            __doPostblit(cast(E[]) arr);
+        else static if (!is(typeof(arr[0].__xpostblit())) && is(immutable T == immutable U, U))
+            foreach (ref elem; (() @trusted => cast(U[]) arr)())
+                elem.__xpostblit();
+        else
+            foreach (ref elem; arr)
+                elem.__xpostblit();
+    }
+}
```
Move these to `core/internal/array/utils.d` because they should be used by multiple hooks

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,181 @@
+    {
+        static if (__traits(isStaticArray, T) && is(T : E[], E))
+            __doPostblit(cast(E[]) arr);
+        else static if (!is(typeof(arr[0].__xpostblit())) && is(immutable T == immutable U, U))
```
Isn't `!__traits(compiles, arr[0].__xpostblit())` what you're looking for instead of `!is(typeof(...))`?

Second: will this ever be used for `immutable` types? From a purely theoretical standpoint it's wrong, but so is `_d_arraysetcapacityPureNothrow`, yet here we are. My point is let's break the type system as little as we can.

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,181 @@
+
+    alias BlkAttr = GC.BlkAttr;
+
+    auto isshared = is(T == shared);
+    alias Unqual_T = Unqual!(T);
```
You can optimise this: extract `isshared` in `reserve()` and pass it as an argument while instantiating the hook with `Unqual_T` instead. This will reduce the number of `_d_arraysetcapacity` as it would only create a singe one for `shared`, `inout`, `const` qualifiers added to the same type.

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -42,9 +217,15 @@ template _d_arraysetlengthTImpl(Tarr : T[], T)
 
             static if (__traits(isZeroInit, T))
-                ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
+
+
+
+                    ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
```
```suggestion
                ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
```

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -42,9 +217,15 @@ template _d_arraysetlengthTImpl(Tarr : T[], T)
+                    ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
             else
-                ._d_arraysetlengthiT(ti, newlength, cast(void[]*)&arr);
+
+
+
+                    ._d_arraysetlengthiT(ti, newlength, cast(void[]*)&arr);
```
```suggestion
                ._d_arraysetlengthiT(ti, newlength, cast(void[]*)&arr);
```
Y tho?

teodutu reviewed (changes requested):
Nice work overall! The failing tests seem unrelated to your work: a timeout and two file-related errors.

Albert24GG committed:
refactor: Move `__doPostblit` and `__typAttrs` to utils

Albert24GG committed:
perf: Instantiate `_d_arraysetcapacity` once for the same base type

Albert24GG committed:
Fix formatting

Albert24GG committed:
Move `__doPostblit` to `core/lifetime.d`

Albert24GG committed:
Update static branch in `__doPostblit`

The previous implementation included a more convoluted way of achieving
the same goal. `is(typeof(arr[0].__xpostblit()))` was replaced by
`__traits(compiles, arr[0].__xpostblit)` and `is(immutable T == immutable
U, U)` with `alias Unqual_T = Unqual!T`

Albert24GG committed:
Improve readability by using aliases

teodutu commented on druntime/src/core/internal/lifetime.d:
```diff
@@ -202,3 +202,22 @@ void swap(T)(ref T lhs, ref T rhs)
+            foreach (ref elem; arr)
+                elem.__xpostblit();
+    }
+}
```
```suggestion
}

```
That's why your pre-commit check fails.

teodutu commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,141 @@
+    size_t gc_reserveArrayCapacity(void[] slice, size_t request, bool atomic) nothrow pure;
+    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) nothrow pure;
+}
+
+
```
```suggestion
```

teodutu reviewed (approved):
I only made some spacing suggestions. Otherwise this looks good. Thanks!

Albert24GG committed:
Fix spacing

ibuclaw commented on druntime/src/core/internal/array/capacity.d:
```diff
@@ -9,6 +9,140 @@
+
+Returns: the number of elements that can actually be stored once the resizing is done
+*/
+size_t _d_arraysetcapacityPureNothrow(T)(size_t newcapacity, void[]* p, bool isshared) pure nothrow @weak @trusted
```
Why is this `@weak`?

Albert24GG replied:
I think that attribute has been preserved from the previous implementation that used extern(C) functions. I think that now with templates, it indeed doesn't make much sense

ibuclaw replied:
Correct. Templates have weak semantics.

Albert24GG committed:
fix: Fix condition for NO_SCAN type attribute

Albert24GG committed:
perf: Store TypeInfo in a static enum variable

This ensures that the TypeInfo will be computed only once, during the
first call made to the templated function instance.

Albert24GG committed:
Remove unnecessary @weak attribute

Albert24GG committed:
fix: Manually check for `__xdtor` instead of `hasElaborateDestructor`

Albert24GG committed:
Remove unused import

teodutu commented:
> @teodutu let me know when this is good to go

It's OK now. The tests are failing because installing the host DMD returns 403, which is unrelated.

Albert24GG committed:
Update `__typeAttrs` to use `hasElaborateDestructor`

teodutu reviewed (approved):
All good from me.

teodutu commented:
@Albert24GG see this comment and create a PR with a changelog entry for this hook as well. Let's keep this in mind for the future.

[1] https://github.com/dlang/dmd/pull/21151#issuecomment-2845670445

CyberShadow commented:
This pull request introduced a regression:
https://github.com/dlang/dmd/issues/23102
